ui/raidboss/raidemulator.js

```diff
@@ -0,0 +1,832 @@
+// the log player when needed.
+class EmulatorView {
+  constructor(
+      fightListElement, timerElement, elapsedElement, infoElement,
+      partyElement, currentPlayerElement, triggerInfoElement
+  ) {
+    this.fightListElement = fightListElement;
+    this.timerElement = timerElement;
+    this.elapsedElement = elapsedElement;
+    this.infoElement = infoElement;
+
+    this.logPlayer = new LogPlayer();
+    this.fightMap = {};
+    this.selectedFight = null;
+    this.startTimeLocalMs = null;
+
+    this.partyElement = partyElement;
+    this.currentPlayerElement = currentPlayerElement;
+    this.triggerInfoElement = triggerInfoElement;
+
+    this.players = {};
+    this.sortedPlayers = [];
+    this.logs = [];
+    this.duration = 0;
+    this.suppressedTriggers = {};
+  }
+
+  DateToTimeStr(date) {
+    let pad2 = function(num) {
+      return ('0' + num).slice(-2);
+    };
+    return pad2(date.getHours()) + ':' + pad2(date.getMinutes());
+  }
+
+  AddFight(fight) {
+    // Note: this uses radio inputs to allow the user to select from a list of
+    // fights that were imported.  It might seem like a <select> would be
+    // more natural, but the native UI for a <select> behaves very badly inside
+    // of CEF, sometimes causing the overlay to be resized (?!).
+    let parentDiv = document.createElement('div');
+    parentDiv.classList.add('fight-option');
+    let radioElement = document.createElement('input');
+    let fightKey = 'fight' + fight.key;
+    this.fightMap[fightKey] = fight;
+
+    radioElement.id = fightKey;
+    radioElement.name = 'fight';
+    radioElement.type = 'radio';
+    let labelElement = document.createElement('label');
+    labelElement.setAttribute('for', radioElement.id);
+
+    let dateStr = this.DateToTimeStr(fight.startDate);
+    let durTotalSeconds = Math.ceil(fight.durationMs / 1000);
+    let durMinutes = Math.floor(durTotalSeconds / 60);
+    let durStr = '';
+    if (durMinutes > 0)
+      durStr += durMinutes + 'm';
+    durStr += (durTotalSeconds % 60) + 's';
+    labelElement.innerText = fight.zoneName + ', ' + dateStr + ', ' + durStr;
+
+    parentDiv.appendChild(radioElement);
+    parentDiv.appendChild(labelElement);
+    this.fightListElement.appendChild(parentDiv);
+
+    radioElement.addEventListener('change', (function() {
+      this.SelectFight(radioElement.id);
+    }).bind(this));
+  }
+
+  SelectFight(fightKey) {
+    let fight = this.fightMap[fightKey];
+    this.selectedFight = fight;
+    localStorage.setItem('fightKey', fightKey);
+
+    if (!this.playingFight) {
+      this.logPlayer.SendZoneEvent(fight.zoneName);
+      this.ShowFightInfo(this.selectedFight);
+
+      // Simply push first Player
+      let name = this.sortedPlayers[0].name;
+      let job = this.sortedPlayers[0].job;
+      let id = this.sortedPlayers[0].id;
+      this.logPlayer.SendPlayerEvent(name, job, id);
+
+      this.Analyze();

```

Comment:
I think this is extremely slow to run this inside of SelectFight.  I think if you want to do the analysis synchronously, it would be better to do that when you hit play, and not when you just click on a fight in the list.  I think you could maybe make this asynchronous if you divided up the work to do and did it in a series of callbacks, but that's definitely more complicated.